src/QsCompiler/Tests.Compiler/CompletionParsingTests.fs

```diff
@@ -0,0 +1,166 @@
+    test "function " [Declaration]
+    test "function Foo" [Declaration]
+    test "function Foo " []
+    test "function Foo (" [Declaration]
+    test "function Foo (x" [Declaration]
+    test "function Foo (x :" [Type]
+    test "function Foo (x : Int" [Type]
+    test "function Foo (x : Int)" []
+    test "function Foo (x : Int) :" [Type]
+    test "function Foo (x : Int) : " [Type]
+    test "function Foo (x : Int) : Unit" [Type]
+    test "function Foo (x : Int) : Unit " []
+    test "function Foo (x : Int) : (Int, MyT" [Type]
+    test "function Foo (x : Int) : (Int," [Type]
+    test "function Foo (x : Int) : (Int, MyT " []
+    test "function Foo (x : (" [Type]
+    test "function Foo (x : (Int" [Type]
+    test "function Foo (x : (Int," [Type]
+    test "function Foo (x : (Int, " [Type]
+    test "function Foo (x : (Int, Int" [Type]
+    test "function Foo (x : (Int, Int " []
+    test "function Foo (x : (Int, Int), " [Declaration]
+    test "function Foo (f : (Int -> " [Type]
+    test "function Foo (f : (Int -> Int)) : Unit" [Type]
+    test "function Foo (f : (Int -> Int)) : Unit " []
+    test "function Foo (q : ((Int -> Int) -> " [Type]
+    test "function Foo (q : ((Int -> Int) -> Int " []
+    test "function Foo (q : ((Int -> Int) -> Int)) : Unit" [Type]
+    test "function Foo (q : ((Int->Int)->Int)) : Unit" [Type]
+    test "function Foo (q : ((Int -> Int) -> Int)) : Unit" [Type]
+    test "function Foo<" [Declaration]
+    test "function Foo<'" [Declaration]
+    test "function Foo<> (q : 'T) : Unit" [Type]
+    test "function Foo<'T> (q : 'T) : Unit" [Type]
+    test "function Foo<'T> (q : ('T -> Int)) : Unit" [Type]
+    test "function Foo<'T> (q : ('T->Int)) : Unit" [Type]
+    test "function Foo<'T> (q : (MyType -> Int)) : Unit" [Type]
+
+[<Fact>]
+let ``Operation declaration parser tests`` () =
+    test "operation " [Declaration]
+    test "operation Foo" [Declaration]
+    test "operation Foo " []
+    test "operation Foo (" [Declaration]
+    test "operation Foo (q" [Declaration]
+    test "operation Foo (q :" [Type]
+    test "operation Foo (q : Qubit" [Type]
+    test "operation Foo (q : Qubit)" []
+    test "operation Foo (q : Qubit) :" [Type]
+    test "operation Foo (q : Qubit) : " [Type]
+    test "operation Foo (q : Qubit) : Unit" [Type]
+    test "operation Foo (q : Qubit) : Unit " [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit i" [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit is" [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit is " [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is A" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj " []
+    test "operation Foo (q : Qubit) : Unit is Adj +" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj + " [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj + Ctl" [Characteristic]
```

Comment:
```suggestion
    test "operation Foo (q : Qubit) : Unit is Adj + Ctl" [Characteristic]
    test "operation Foo (q : Qubit) : Unit is (" [Characteristic]
    test "operation Foo (q : Qubit) : Unit is ( " [Characteristic]
    test "operation Foo (q : Qubit) : Unit is (Adj" [Characteristic]
    test "operation Foo (q : Qubit) : Unit is (Adj " []
    test "operation Foo (q : Qubit) : Unit is (Adj + " [Characteristic]
    test "operation Foo (q : Qubit) : Unit is (Adj + Ctl" [Characteristic]
    test "operation Foo (q : Qubit) : Unit is (Adj + Ctl)" []    
```

Reply:
Some of these tests fail because the parser I'm not using a custom characteristics expression parser in this PR. It ends up working for simple cases but gets confused by missing parentheses and will think that `operation Foo (q : Qubit) : Unit is (Adj ` should be followed by another characteristic.

In my expression parser branch, I have not tried adapting the new expression parser for characteristics yet, but I think it will fix these issues. Is it OK to wait until a later PR?